src/main/java/htsjdk/samtools/BAMFileReader.java

```diff
@@ -407,20 +407,28 @@ public boolean hasIndex() {
      */
     @Override
     public BAMIndex getIndex() {
-        if(!hasIndex())
+        if(!hasIndex()) {
             throw new SAMException("No index is available for this BAM file.");
+        }
         if(mIndex == null) {
-            SamIndexes samIndex = getIndexType();
-            if (samIndex == null) {
-                mIndex = mEnableIndexCaching ? new CachingBAMFileIndex(mIndexStream, getFileHeader().getSequenceDictionary())
-                        : new DiskBasedBAMFileIndex(mIndexStream, getFileHeader().getSequenceDictionary());
-            } else if (samIndex.equals(SamIndexes.BAI)) {
+            final SamIndexes samIndex = getIndexType();
+            if(mIndexFile != null) {
+                if (samIndex.equals(SamIndexes.BAI)) {
                     mIndex = mEnableIndexCaching ? new CachingBAMFileIndex(mIndexFile, getFileHeader().getSequenceDictionary(), mEnableIndexMemoryMapping)
```

Comment:
we might pull out a sequenceDictionary local variable to condense the 5 invocations of getFileHeader().getSequenceDictionary()

Reply:
Done.